src/graph_world/generators/lfr_simulator.py

```diff
@@ -0,0 +1,204 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import collections
+import enum
+import math
+import random
+from typing import Dict, Sequence, List, Tuple
+from sklearn.preprocessing import normalize
+
+import dataclasses
+import graph_tool
+import numpy as np
+import networkit as nk
+
+from graph_tool.all import *
+from graph_world.generators.sbm_simulator import SimulateFeatures, MatchType, SimulateEdgeFeatures
+
+@dataclasses.dataclass
+class LFR:
+  """
+  Stores data for the LFR Model.
+  """
+  graph: graph_tool.Graph = Ellipsis 
+  graph_memberships: np.ndarray = Ellipsis 
+  node_features: np.ndarray = Ellipsis 
+  feature_memberships: np.ndarray = Ellipsis 
+  edge_features: Dict[Tuple[int, int], np.ndarray] = Ellipsis 
+
+def NetworkitToGraphWorldData(G):
+  """
+  Converts NetworKit graph data to GraphWorld LFR dataclass, with node memberships.
+  Args:
+    G: Networkit generated LFR graph
+  Returns:
+    lfr_gt: LFR dataclass instance to store graph data
+  """
+  nk_edges = list(G.iterEdges())
+  nk_nodes = list(G.iterNodes())
+  lfr_gt = graph_tool.Graph(directed=False)
+  vertices = {} 
+  for node in nk_nodes:
+      v = lfr_gt.add_vertex()
+      vertices[node] = v
+  for src, dst in nk_edges:
+      e = lfr_gt.add_edge(vertices[src], vertices[dst])
+  return lfr_gt
+
+def SimulateLFR(n,
+                avg_deg,
+                max_deg,
+                exponent,
+                min_community_size,
+                max_community_size,
+                exponent_community,
+                mu):
+  """
+  Simulates an LFR Graph using NetworKit and the sampled parameters. 
+  Args:
+    n: number of nodes 
+    avg_deg: average degree 
+    max_deg: maximum degree 
+    exponent: power law exponent that the degree sequence should follow
+    min_community_size: minimum community size 
+    max_community_size: maximum community size 
+    exponent_community: power law exponent that the community sizes should follow
+    mu: mixing parameter
+  Returns:
+    lfrG: generated NetworKit LFR graph
+    community_sizes: sequence of community sizes in generated graph
+  """
+  lfr = nk.generators.LFRGenerator(n)
+  lfr.generatePowerlawDegreeSequence(avg_deg, 
```

Comment:
Please adjust these indentations -- put on the same line if possible, align indentations if not